mantidproject/mantid
# Crash when opening plot options on a sample logs plot

**Description of work.**

This fixes two issues where the solutions are intertwined.
Both affect sample log plots when they have been popped out of the sample log dialog into separate plots:
1. Plot options & the right-click menu crash
2. Scripts generated from the plot do not recreate the original graph (they error)

Aspects of the work:
1. Store log plotting details into the artist class so they can be accessed later.
2. When generating scripts extract the log name, Filtered and expt info to populate the generation string.
3. Add protection from None being passed to C++ when trying to determine the ws index for a spectra number (log plots don't have one).
4. Enable recreation of the plot after changes in the plot options dialog.  This required a few limited changes where it was expected that the data should have errors, but doesn't for sample logs.
5. Hide the "Fit" button is the plot contains a sample log plot.



**To test:**

Test the plot options dialog for a sample log plot:

1. Load some data
   ``` python
   ts_logs = Load("ENGINX00228061")
   ```
1. Right-click and select Show Sample Logs
1. Scroll down and click any float series log.
1. Double-click the plot so it pops out.
1. Click the cog in the toolbar.
1. Make changes in the plot options dialog, make sure they are made to the graph.
1. Also check the right-click context menu works:
   1. For sample log plots
   1. For "normal plots"
       1. One issue I found was that a normal plot may be rescaled after changing some aspect of the curve (line the line color).  It looks like it is being rescaled to make room for error bars that are not shown.  However, this is already present on master and is not associated with this PR.

Then test script generation:

1. Open Workbench load a file with some logs ``` ts_logs = Load("ENGINX00228061")```
1. Workspace->show sample logs
1. pick a log with a plot you would recognize (w has a square wave pattern)
1. Double click on the plot to open it externally
1. There should not be a fit button visible at the top of the plot window
1. on the created plot click the Script icon (don't click the cog or right-click as this will crash - that is a different issue)
1. Paste the script into a python window and execute it to check it recreates the same plot.
1. Now check that other plot types that could be scripted still work.


This script also quickly loads and plots a sample log plot
``` python
ts_logs = Load("ENGINX00228061")

fig, axes = plt.subplots(edgecolor='#ffffff', num='Figure 1', subplot_kw={'projection': 'mantid'})
axes.plot(ts_logs, ExperimentInfo=0, Filtered=True, LogName='w', color='#1f77b4', drawstyle='steps-post', label='w')
axes.set_xlabel('Time (s)')
axes.legend().set_draggable(True)

fig.show()
```

Fixes #28635

Fixes #27770


---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
